Add "reference" as a known compiletest header

ehuss opened the pull request:
This adds the "reference" compiletest header so that the Rust reference can add annotations to the test suite in order to link tests to individual rules in the reference.

Tooling in the reference repo will be responsible for collecting these annotations and linking to the tests.

More details are in MCP 783: https://github.com/rust-lang/compiler-team/issues/783

There is a change from the MCP in that I am not adding the JSON collection to compiletest (at least, not yet). In looking at this more closely, that actually makes things more difficult for our tooling, so I'm leaving it out for now. If in the future it looks like something we want, then I think we can add it later.

There are a few tests here which need adjusting due to the legacy header check. @jieyouxu indicated on Zulip that we could potentially remove the legacy header check, in which case those changes can be dropped from this PR.

r? @jieyouxu

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11224974995/job/31202782374) [(plain)](https://github.com/rust-lang/rust/commit/fe3385691fd8917e7f227c02f468b20d8a2c4f7b/checks/31202782374/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.851 Building wheels for collected packages: reuse
#13 2.852   Building wheel for reuse (pyproject.toml): started
#13 3.095   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.096   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.097   Stored in directory: /tmp/pip-ephem-wheel-cache-j8h_6q1r/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.099 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.491 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.491 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.017 Collecting virtualenv
#13 4.017 Collecting virtualenv
#13 4.067   Downloading virtualenv-20.26.6-py3-none-any.whl (6.0 MB)
#13 4.234      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 36.4 MB/s eta 0:00:00
#13 4.295 Collecting filelock<4,>=3.12.2
#13 4.302   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.339 Collecting platformdirs<5,>=3.9.1
#13 4.348   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.368 Collecting distlib<1,>=0.3.7
#13 4.386      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 61.8 MB/s eta 0:00:00
#13 4.386      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 61.8 MB/s eta 0:00:00
#13 4.466 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.653 Successfully installed distlib-0.3.8 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.26.6
#13 DONE 4.7s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      214976 kB
DirectMap2M:    11319296 kB
DirectMap1G:     7340032 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
downloading https://static.rust-lang.org/dist/2024-09-22/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/2024-09-22/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt
---
fmt check
fmt: checked 5592 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/tests/run-make/include-all-symbols-linking/rmake.rs:2: line longer than 100 chars
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.2)
All checks passed!
checking C++ file formatting
some tidy checks failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Mon Oct  7 22:33:41 UTC 2024
  network time: Mon, 07 Oct 2024 22:33:41 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

jieyouxu commented:
I'll drop the legacy directive checks before this PR, it's been at least 6 months since we migrated from `//@`.

jieyouxu commented:
PR dropping the legacy directive check: https://github.com/rust-lang/rust/pull/131392

ehuss committed:
Add "reference" as a known compiletest header

This adds the "reference" compiletest header so that the Rust reference
can add annotations to the test suite in order to link tests to
individual rules in the reference.

Tooling in the reference repo will be responsible for collecting these
annotations and linking to the tests.

More details are in MCP 783: https://github.com/rust-lang/compiler-team/issues/783

jieyouxu reviewed (approved):
Thanks! Could you also update dev-guide about a new directive (even if it's a no-op currently).

jieyouxu commented:
@bors r+ rollup

bors commented:
:pushpin: Commit de60931645a1baf5edddd9f0c9027c0e494b46fb has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu de60931645a1baf5edddd9f0c9027c0e494b46fb -->
<!-- homu: {"type":"Approved","sha":"de60931645a1baf5edddd9f0c9027c0e494b46fb","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
